agent/system_metrics.go

```diff
@@ -0,0 +1,245 @@
+	if err != nil {
+		return nil, errors.Wrap(err, "error getting cedar dial options")
+	}
+	dialOpts := timber.DialCedarOptions{
+		BaseAddress: bi.BaseURL,
+		RPCPort:     bi.RPCPort,
+		Username:    bi.Username,
+		Password:    bi.Password,
+		APIKey:      bi.APIKey,
+		Retries:     10,
+	}
+	connOpts := system_metrics.ConnectionOptions{
+		DialOpts: dialOpts,
+	}
+
+	return &SystemMetricsCollector{
+		logger:     logger,
+		interval:   interval,
+		collectors: collectors,
+		taskOpts:   taskOpts,
+		connOpts:   connOpts,
+	}, nil
+}
+
+func (s *SystemMetricsCollector) Start(ctx context.Context) {
+	catcher := grip.NewBasicCatcher()
+	defer func() {
+		s.mu.Lock()
+		defer s.mu.Unlock()
+		s.errored = true
+		err := catcher.Resolve()
+		s.logger.System().Error(message.WrapError(err, message.Fields{
+			"message":  "error setting up system metrics collection",
+			"id":       s.id,
+			"connOpts": s.connOpts,
+			"taskOpts": *s.taskOpts,
+		}))
+	}()
+
+	var err error
+	s.client, err = system_metrics.NewSystemMetricsClient(ctx, s.connOpts)
+	if err != nil {
+		catcher.Add(err)
+		return
+	}
+
+	s.id, err = s.client.CreateSystemMetricRecord(ctx, *s.taskOpts)
+	if err != nil {
+		catcher.Add(err)
+		return
+	}
+
+	innerCtx := context.Background()
+	innerCtx, s.innerCancel = context.WithCancel(innerCtx)
+
+	for _, collector := range s.collectors {
+		s.wg.Add(1)
+		stream, err := s.client.StreamSystemMetrics(innerCtx, s.id, collector.Name(), system_metrics.StreamOpts{})
+		if err != nil {
+			catcher.Add(err)
+			return
+		}
+		ftdcCollector := ftdc.NewStreamingCollector(1, stream)
+		err = ftdcCollector.SetMetadata(collector.Metadata())
+		if err != nil {
+			catcher.Add(err)
+			return
+		}
+		go s.timedCollect(innerCtx, collector, stream, ftdcCollector)
+	}
+	go s.closeOnCancel(ctx)
+}
+
+func (s *SystemMetricsCollector) timedCollect(ctx context.Context, mc MetricCollector, stream *system_metrics.SystemMetricsWriteCloser, collector ftdc.Collector) {
+	defer s.wg.Done()
+
+	catcher := grip.NewBasicCatcher()
+	defer func() {
+		s.mu.Lock()
+		defer s.mu.Unlock()
+		s.errored = true
+		err := catcher.Resolve()
+		s.logger.System().Error(message.WrapError(err, message.Fields{
+			"message":    "error while collecting system metrics",
+			"id":         s.id,
+			"metricType": mc.Name(),
+		}))
+	}()
+
+	timer := time.NewTimer(0)
+	defer timer.Stop()
+	defer catcher.Add(stream.Close())
+	defer ftdc.FlushCollector(collector, stream)
+	defer catcher.Add(recovery.HandlePanicWithError(recover(), nil, "panic in system metrics collector"))
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-timer.C:
+			func() {
+				data, err := mc.Collect()
+				if err != nil {
+					catcher.Add(err)
+					return
+				}
+				err = collector.Add(data)
+				if err != nil {
+					catcher.Add(err)
+					return
+				}
+				_ = timer.Reset(s.interval)
+			}()
+		}
+	}
+}
+
+func (s *SystemMetricsCollector) closeOnCancel(outerCtx context.Context) {
```

Comment:
outerCtx -> ctx

Usually you only name contexts different from ctx if you have multiple contexts.